inc/Utf8/Sort.php

```diff
@@ -0,0 +1,154 @@
+<?php
+
+namespace dokuwiki\Utf8;
+
+/**
+ * DokuWiki sort functions
+ *
+ * When "intl" extension is available, all sorts are done using a collator.
+ * Otherwise, primitive PHP functions are called.
+ *
+ * The collator is created using the locale given in $conf['lang'].
+ * It always uses case insensitive "natural" ordering in its collation.
+ * The fallback solution uses the primitive PHP functions that return almost the same results
+ * when the input is text with only [A-Za-z0-9] characters.
+ *
+ * @license    GPL 2 (http://www.gnu.org/licenses/gpl.html)
+ * @author     Moisés Braga Ribeiro <moisesbr@gmail.com>
+ */
+class Sort
+{
+    /** @var \Collator[] language specific collators, usually only one */
+    protected static $collator;
+
+    /**
+     * Initialization of a collator using $conf['lang'] as the locale.
+     * The initialization is done only once, except when $reload is set to true.
+     * The collation takes "natural ordering" into account, that is, "page 2" is before "page 10".
+     *
+     * @param bool $reload Usually false; true forces collator re-creation
+     * @return \Collator Returns a configured collator or null if the collator cannot be created.
+     *
+     * @author Moisés Braga Ribeiro <moisesbr@gmail.com>
+     */
+    protected static function getCollator($reload = false)
+    {
+        global $conf;
+        $lc = $conf['lang'];
+
+        // check if intl extension is available
+        if (!class_exists('\Collator')) {
+            return null;
+        }
+
+        // load collator if not available yet
+        if ($reload || !isset(self::$collator[$lc])) {
+            $collator = \Collator::create($lc);
+            if (!isset($collator)) return null;
+            $collator->setAttribute(\Collator::NUMERIC_COLLATION, \Collator::ON);
```

Comment:
Please reinsert the dbglog() call after this line. It would be used to check the valid and actual locales of the created collator, in case of problems. Some language codes yield the default collator, which may give wrong results.

            dbglog('Collator created with locale "' . $lc . '": numeric collation on, ' .
                   'valid locale "' . $collator->getLocale(Locale::VALID_LOCALE) . '", ' .
                   'actual locale "' . $collator->getLocale(Locale::ACTUAL_LOCALE) . '"');

Reply:
I am unsure about this. These kind of debug statements are often useful during development, but nearly never needed again later on and clutter the logs unnecessary. But if you feel strongly about it, keep it in.

Reply:
[This comment](https://github.com/splitbrain/dokuwiki/issues/2037#issuecomment-623194166) in issue #2037 explains the whole situation. Please download and run the tests given there, of course changing the file extension.